lib/charms/prometheus_pushgateway_k8s/v0/pushgateway.py

```diff
@@ -0,0 +1,172 @@
+
+    def _on_relation_changed(self, event: RelationEvent):
+        """Send the push endpoint info."""
+        relation_data = event.relation.data[self.app]
+        relation_data[RELATION_KEY] = json.dumps(
+            {
+                "hostname": socket.getfqdn(),
+                "port": self.port,
+            }
+        )
+
+
+class PrometheusPushgatewayRequirer(Object):
+    """Requirer side for the Prometheus Pushgateway."""
+
+    def __init__(self, charm: CharmBase, relation_name: str):
+        """Construct the interface for the Prometheus Pushgateway.
+
+        Args:
+            charm: a `CharmBase` object that manages this object. Typically,
+                this is `self` in the instantiating class.
+            relation_name: the name of the relation (whatever was used
+                in the `requires` section in `metadata.yaml` for
+                the `pushgateway` interface.
+        """
+        super().__init__(charm, relation_name)
+        self._relation_name = relation_name
+
+    @property
+    def _pushgateway_url(self) -> Optional[str]:
+        """Build the pushgateway url using the relation data (if present, else return None)."""
+        relation = self.model.get_relation(self._relation_name)
+        if relation is None:
+            logger.warning(
+                "Prometheus Pushgateway Requirer not ready: "
+                "charm not related to the Pushgateway service")
+            return None
+        raw_data = relation.data[relation.app].get(RELATION_KEY)
+        if raw_data is None:
+            logger.warning(
+                "Prometheus Pushgateway Requirer not ready: still no data in the relation")
+            return None
+        data = json.loads(raw_data)
+        return "http://{hostname}:{port}/".format_map(data)
+
+    def is_ready(self):
+        """Return if the service is ready to send metrics."""
+        return self._pushgateway_url is not None
+
+    def send_metric(self, name: str, value: Union[float, int], ignore_error: bool = False):
+        """Send a metric to the Pushgateway."""
+        # This currently follows the "simple API" for the case of one metric
+        # without labels, as indicated here:
+        #    https://github.com/prometheus/pushgateway#api
+        # TODO: support the more complex cases
+
+        pushgateway_url = self._pushgateway_url
+        if pushgateway_url is None:
+            raise ValueError("The service is not ready.")
+        if not isinstance(name, str) or not name.isascii() or not name:
+            raise ValueError("The name must be a non-empty ASCII string.")
+        if not isinstance(value, (float, int)):
+            raise ValueError("The metric value must be an integer or float number.")
+
+        payload = f"{name} {value}\n".encode("ascii")
+        post_url = pushgateway_url + "metrics/job/testjob"
+
+        try:
+            request.urlopen(post_url, data=payload)
```

Comment:
Are we ok here to keep the default timeouts?
If a charm attempts multiple `send_metric` in the same hook then juju may terminate it eventually if the timeout is too long?

Reply:
The Pushgateway service responded very fast in the tests I did, but of course real cloud deployments may be different.

The problem with using a very small timeout is that if the service is kind of slow and the charm sends only one metric, it could fail, and it could have waited a little more.

Probably the best approach would be to consider adding a timeout parameter in the function itself, so the charm developer using this library could have control on which timeout to have (and it should be sensible for the most common case of one or two metrics, but the user could tune it for corner cases).

All that said, the lower the timeout, the most probable to hit the situation of "receiving timeout this side of the wire but the metric being recorded on the other side" (which would lead to repeated metrics).

For sure something to consider for the future, I don't think we should define for this PR.